critical_demand.py

```diff
@@ -35,513 +35,892 @@
+        ]
+    else:
+        capacity_inverter = 0
+
+    if "scalars" in results_rectifier:
+        capacity_rectifier = results_rectifier["scalars"][
+            (("electricity_ac", "rectifier"), "invest")
+        ]
+    else:
+        capacity_rectifier = 0
+
+    df_results.loc["diesel_genset", "capacity"] = capacity_diesel_genset
+    df_results.loc["pv", "capacity"] = capacity_pv
+    df_results.loc["battery", "capacity"] = capacity_battery
+    df_results.loc["inverter", "capacity"] = capacity_inverter
+    df_results.loc["rectifier", "capacity"] = capacity_rectifier
+
+    # Scaling annuity to timeframe
+    year_fraction = n_days / n_days_in_year
+
+    # Compute annual costs for each components
+    df_results["annual_costs"] = df_results.apply(
+        lambda x: (x.annuity * x.capacity) * year_fraction
+        + x.total_flow * x.opex_variable,
+        axis=1,
+    )
 
-results_pv = solph.views.node(results=results, node="pv")
-results_diesel_source = solph.views.node(results=results, node="diesel_source")
-results_diesel_genset = solph.views.node(results=results, node="diesel_genset")
-results_inverter = solph.views.node(results=results, node="inverter")
-results_rectifier = solph.views.node(results=results, node="rectifier")
-if case != case_BPV:
-    results_battery = solph.views.node(results=results, node="battery")
-else:
-    results_battery = None
+    df_results["total_opex_costs"] = df_results.apply(
+        lambda x: (x.opex_fix * x.capacity) * year_fraction
+        + x.total_flow * x.opex_variable,
+        axis=1,
+    )
 
-results_demand_el = solph.views.node(
-    results=results, node="electricity_demand"
-)
-results_excess_el = solph.views.node(results=results, node="excess_el")
-
-# results_critical_demand_el = ?
-# TODO also fetch the results of the critical demand
-
-# -------------------- SEQUENCES (DYNAMIC) --------------------
-# Hourly demand profile.
-sequences_demand = results_demand_el["sequences"][
-    (("electricity_ac", "electricity_demand"), "flow")
-]
-
-# Hourly profiles for solar potential and pv production.
-sequences_pv = results_pv["sequences"][(("pv", "electricity_dc"), "flow")]
-
-# Hourly profiles for diesel consumption and electricity production
-# in the diesel genset.
-# The 'flow' from oemof is in kWh and must be converted to
-# kg by dividing it by the lower heating value and then to
-# liter by dividing it by the diesel density.
-sequences_diesel_consumption = (
-    results_diesel_source["sequences"][(("diesel_source", "diesel"), "flow")]
-    / diesel_lhv
-    / diesel_density
-)
+    # Save the results
+    df_results = df_results[
+        [
+            "annuity",
+            "annual_costs",
+            "total_flow",
+            "capacity",
+            "cash_flow",
+            "total_opex_costs",
+        ]
+    ]
+    df_results.to_csv(f"results_{case}.csv")
+
+    NPV = (df_results.annual_costs.sum() + df_results.cash_flow.sum()) / CRF
```

Comment:
Hi @Bachibouzouk PF, I guess `annual_costs` and `cash_flow` are not defined yet. Also, `first_investment` KPI for the three cases is also missed.

Reply:
@adnanalakori  If you can run the code without error that means that `annual_costs` and `cash_flow` are defined somewhere above this line. Use `ctrl + f` to look occurences of `annual_costs` (name of the column in the dataframe `df_results`) or `cash_flow`

Reply:
Thanks, I figured its value through looking for whole variable  `df_results.annual_costs.sum`, before I was debugging  `annual_costs` or `cash_flow` which shows as an empty (non defined) string one.